src/hooks/appPage.ts

```diff
@@ -0,0 +1,189 @@
+  >(
+    gql`
+      query GET_APP_PAGE_COLLECTION($appId: String!) {
+        app_page(where: { app_id: { _eq: $appId }, is_deleted: { _eq: false } }) {
+          id
+          path
+          title
+          craft_data
+          updated_at
+          published_at
+          options
+          editor_id
+          editor {
+            name
+          }
+        }
+      }
+    `,
+    {
+      variables: {
+        appId,
+      },
+    },
+  )
+
+  const appPages: AppPageProps[] =
+    data?.app_page
+      .map(v => ({
+        id: v.id,
+        path: v.path,
+        title: v.title,
+        craftData: v.craft_data,
+        editorId: v.editor_id,
+        editorName: v.editor?.name || null,
+        publishedAt: v.published_at ? new Date(v.published_at) : null,
+        updatedAt: v.updated_at ? new Date(v.updated_at) : null,
+        options: v.options,
+      }))
+      .filter(v => v.craftData) || []
+
+  return {
+    refetchAppPages: refetch,
+    loadingAppPages: loading,
+    errorAppPages: error,
+    appPages,
+  }
+}
+
+export const useMutateAppPage = () => {
+  const { id: appId } = useApp()
+  const [insertAppPageHandler] = useMutation<hasura.INSERT_APP_PAGE, hasura.INSERT_APP_PAGEVariables>(gql`
+    mutation INSERT_APP_PAGE(
+      $appId: String!
+      $path: String
+      $editorId: String!
+      $title: String!
+      $craftData: jsonb
+      $options: jsonb
+    ) {
+      insert_app_page_one(
+        object: {
+          app_id: $appId
+          path: $path
+          title: $title
+          editor_id: $editorId
+          craft_data: $craftData
+          options: $options
+        }
+      ) {
+        id
+      }
+    }
+  `)
+  const insertAppPage = (values: {
+    path?: string
+    title: string
+    editorId: string
+    craftData: { [key: string]: any } | null
+  }) => {
+    return insertAppPageHandler({
+      variables: { appId, ...values, options: { white: true } },
+    })
+  }
+  const [updateAppPageHandler] = useMutation<hasura.UPDATE_APP_PAGE, hasura.UPDATE_APP_PAGEVariables>(gql`
+    mutation UPDATE_APP_PAGE($pageId: uuid!, $updated: app_page_set_input) {
+      update_app_page_by_pk(pk_columns: { id: $pageId }, _set: $updated) {
+        id
+      }
+    }
+  `)
+  const updateAppPage = (values: {
+    pageId: string
+    path?: string
+    title?: string
+    editorId?: string
+    publishedAt?: Date | null
+    craftData?: { [key: string]: any }
+    isDeleted?: boolean
+  }) => {
+    return updateAppPageHandler({
+      variables: {
+        pageId: values.pageId,
+        updated: {
+          ...(values.path ? { path: values.path } : {}),
+          ...(values.title ? { title: values.title } : {}),
+          ...(values.editorId ? { editor_id: values.editorId } : {}),
+          ...(values.publishedAt !== undefined ? { published_at: values.publishedAt } : {}),
+          ...(values.craftData ? { craft_data: values.craftData } : {}),
+          is_deleted: values.isDeleted,
+        },
```

Comment:
nice pattern to avoid mutating other columns!

Reply:
```js
const removeUndefinedValue = obj => JSON.parse(JSON.stringify(obj))
```

Reply:
```graphql
mutation UPDATE_APP(_set: $updated) {
}
```

```js
{
   variables: {
      updated: {
          title: 'foo'
          description: undefined
      }
   }
}
```

```js
const { update } = useProgram(programId)
update({ title: "xxx" })
```